cmd/knitgateway/main.go

```diff
@@ -88,19 +92,88 @@ func main() {
 
 	mux := http.NewServeMux()
 	mux.Handle(gateway.AsHandler())
-	corsHandler := cors.AllowAll().Handler(mux)
+	certCapturingHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.TLS != nil && len(r.TLS.PeerCertificates) > 0 {
+			r = r.WithContext(knitgateway.ContextWithClientCert(r.Context(), r.TLS.PeerCertificates[0]))
+		}
+		mux.ServeHTTP(w, r)
+	})
+	// TODO: CORS should not be so lenient unless configured to be so
+	corsHandler := cors.AllowAll().Handler(certCapturingHandler)
 	loggingHandler := knitgateway.NewLoggingHandler(corsHandler, logger)
 	svr := http.Server{
 		Handler:           h2c.NewHandler(loggingHandler, &http2.Server{}),
 		ReadHeaderTimeout: 30 * time.Second,
+		TLSConfig:         config.TLSConfig,
 	}
-	l, err := net.Listen("tcp", config.ListenAddress)
-	if err != nil {
-		fatalf("failed to listen on bind address %s: %v\n", config.ListenAddress, err)
+
+	ignored := make(chan os.Signal, 1)
+	signal.Notify(ignored, syscall.SIGHUP)
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, syscall.SIGTERM, syscall.SIGINT, syscall.SIGALRM, syscall.SIGQUIT, syscall.SIGTSTP, syscall.SIGUSR1, syscall.SIGUSR2)
+
+	group, grpCtx := errgroup.WithContext(ctx)
+	group.Go(func() error {
+		// if either listener below stops or we get a signal, shut down the whole server
+		select {
+		case <-grpCtx.Done():
+		case s := <-sig:
+			logger.Sugar().Infof("Received signal %v. Shutting down...", s)
+		}
+		defer signal.Stop(sig)
+
+		timeoutCtx, cancel := context.WithTimeout(ctx, 30*time.Second)
+		defer cancel()
+		if err := svr.Shutdown(timeoutCtx); err != nil {
+			logger.Sugar().Errorf("could not complete shutdown: %w", err)
+		}
+		return nil
+	})
+
+	scheme := "HTTP"
+	serveFunc := svr.Serve
+	if config.TLSConfig != nil {
+		scheme = "HTTPS"
+		serveFunc = func(l net.Listener) error {
+			return svr.ServeTLS(l, "", "")
+		}
 	}
-	logger.Sugar().Infof("Listening on %s for HTTP requests...", l.Addr().String())
-	if err := svr.Serve(l); err != nil {
-		fatalf("HTTP server failed: %v", err)
+
+	// NB: goroutines below must return non-nil errors in order for grpCtx
```

Comment:
What is a case where server returned and the above goroutine did not?

Reply:
I'm not certain as `http.Server.Serve` doesn't provide all of the cases where it could return prematurely (before the server is shutdown).

Scanning through the implementation, I see two possibilities:
* The listener's `Accept` method returns an error. I'm uncertain what kinds of issues might cause this. Maybe it's more likely with unix sockets than with TCP sockets, like an admin deletes the unix socket from the filesystem out from underneath the application?
* The configured cipher suites are incompatible with HTTP/2. HTTP/2 requires at least one of ECDHE_RSA_WITH_AES_128_GCM_SHA256 or ECDHE_ECDSA_WITH_AES_128_GCM_SHA256.

But since it isn't clearly specified, it seemed possible there could be other failure modes (including ones added in the future). So this is a "just in case" to prevent the gateway from basically freezing (where HTTP server has effectively stopped serving, but binary won't exit).